File: scope/android/app/src/main/kotlin/com/scope/attentions/HashedLogger.kt
package com.scope.attentions

import android.util.Log

/**
 * Centralized platform logging utility for Android.
 *
 * Ensures notification titles and body content are transformed into deterministic
 * DJB2 cryptographic hashes and structural metadata prior to log emission.
 */
object HashedLogger {

    /**
     * Computes deterministic DJB2 hash for a string, matching Dart's DJB2 implementation.
     */
    fun djb2(valStr: String?): Long {
        if (valStr == null) return 0L
        var hash = 5381L
        for (ch in valStr) {
            hash = ((hash shl 5) + hash) + ch.code.toLong()
            hash = hash and 0xFFFFFFFFL
        }
        return hash
    }

    /**
     * Helper to compute character count safely.
     */
    fun charCount(str: String?): Int = str?.length ?: 0

    /**
     * Logs notification capture event with structural metadata and DJB2 string hashes.
     * Replaces cleartext title and body with non-invertible hashes and character counts.
     */
    fun logNotificationCaptured(
        tag: String,
        packageName: String,
        title: String?,
        content: String?,
        category: String? = null,
        isOngoing: Boolean = false
    ) {
        val titleHash = djb2(title ?: "")
        val contentHash = djb2(content ?: "")
        val titleLen = charCount(title)
        val contentLen = charCount(content)
        val hasCategory = !category.isNullOrEmpty()

        Log.d(
            tag,
            "Captured: pkg=$packageName, titleHash=$titleHash (len=$titleLen), contentHash=$contentHash (len=$contentLen), category=$category, isOngoing=$isOngoing, hasCategory=$hasCategory"
        )
    }

    /**
     * Logs notification removal event with structural metadata and DJB2 string hash.
     */
    fun logNotificationRemoved(
        tag: String,
        packageName: String,
        title: String?
    ) {
        val titleHash = djb2(title ?: "")
        val titleLen = charCount(title)
        Log.d(
            tag,
            "Removed: pkg=$packageName, titleHash=$titleHash (len=$titleLen)"
        )
    }

    /** Log wrappers for standard severity levels without payload content */
    fun debug(tag: String, message: String) {
        Log.d(tag, message)
    }

    fun info(tag: String, message: String) {
        Log.i(tag, message)
    }

    fun warn(tag: String, message: String) {
        Log.w(tag, message)
    }

    fun error(tag: String, message: String, throwable: Throwable? = null) {
        if (throwable != null) {
            Log.e(tag, message, throwable)
        } else {
            Log.e(tag, message)
        }
    }
}

File: scope/lib/core/utils/hashed_logger.dart
import 'package:flutter/foundation.dart';
import 'package:scope/core/analysis/ghost_ai.dart';
import 'package:scope/core/models/notification_model.dart';

/// Centralized platform logging abstraction for Flutter.
///
/// Ensures raw notification titles and body content are replaced with
/// deterministic DJB2 hashes and structural metadata prior to log emission.
class HashedLogger {
  /// Computes deterministic DJB2 string hash matching Android native implementation.
  static int djb2(String val) {
    int hash = 5381;
    for (int i = 0; i < val.length; i++) {
      hash = ((hash << 5) + hash) + val.codeUnitAt(i);
      hash = hash & 0xFFFFFFFF;
    }
    return hash;
  }

  /// Helper to get character length safely.
  static int charCount(String? val) => val?.length ?? 0;

  /// Logs Ghost AI inference report with structural metadata and hashes.
  static void logInferenceReport(AppNotification notification, GhostAIResult result) {
    final titleHash = djb2(notification.title);
    final titleLen = notification.title.length;
    final contentHash = djb2(notification.content);
    final contentLen = notification.content.length;
    final hasCategory = notification.category != null && notification.category!.isNotEmpty;

    final features = notification.extractedFeatures ?? {};
    final hasMoney = features['hasMoney'] == true || features['contains_money'] == true;
    final hasOtp = features['hasOtp'] == true || features['contains_otp'] == true;
    final hasLink = features['hasLink'] == true || features['contains_link'] == true;

    debugPrint('=== GHOST AI INFERENCE REPORT ===');
    debugPrint('Notification ID: ${notification.id}');
    debugPrint('Package: ${notification.packageName}');
    debugPrint('Title Hash: $titleHash (length: $titleLen)');
    debugPrint('Content Hash: $contentHash (length: $contentLen)');
    debugPrint('Metadata: category=${notification.category}, hasCategory=$hasCategory, isOngoing=${notification.isOngoing}');
    debugPrint('Entity Flags: hasMoney=$hasMoney, hasOtp=$hasOtp, hasLink=$hasLink');
    debugPrint('Feature Vector (First 15): ${result.featureVector.take(15).toList()}...');
    debugPrint('Inference Time: ${result.inferenceTimeUs} us');
    debugPrint('Raw Predicted Score: ${(result.predictedScore * 100).toStringAsFixed(2)}');
    debugPrint('Rule Score: ${result.ruleScore != null ? (result.ruleScore! * 100).toStringAsFixed(2) : "N/A"}');
    debugPrint('Final Fused Score: ${(result.reviewScore * 100).toStringAsFixed(2)}');
    debugPrint('==================================');
  }

  /// Logs a general notification event with structural metadata and hashes.
  static void logNotification(
    String tag,
    AppNotification notification, {
    String? event,
  }) {
    final titleHash = djb2(notification.title);
    final titleLen = notification.title.length;
    final contentHash = djb2(notification.content);
    final contentLen = notification.content.length;
    final prefix = event != null ? '[$event] ' : '';

    debugPrint(
      '[$tag] ${prefix}pkg=${notification.packageName}, id=${notification.id}, '
      'titleHash=$titleHash (len=$titleLen), contentHash=$contentHash (len=$contentLen)',
    );
  }

  /// Log wrappers for standard severity levels without raw payload text.
  static void logDebug(String tag, String message) {
    debugPrint('[$tag] [DEBUG] $message');
  }

  static void logInfo(String tag, String message) {
    debugPrint('[$tag] [INFO] $message');
  }

  static void logWarning(String tag, String message) {
    debugPrint('[$tag] [WARN] $message');
  }

  static void logError(String tag, String message, [dynamic error, StackTrace? stackTrace]) {
    final errStr = error != null ? ' Error: $error' : '';
    debugPrint('[$tag] [ERROR] $message$errStr');
    if (stackTrace != null) {
      debugPrint(stackTrace.toString());
    }
  }
}
